packages/foam-core/src/model/uri.ts

```diff
@@ -0,0 +1,508 @@
+  }
+  return uri;
+};
+
+export const computeRelativeURI = (
+  reference: URI,
+  relativeSlug: string
+): URI => {
+  // if no extension is provided, use the same extension as the source file
+  const slug =
+    posix.extname(relativeSlug) !== ''
+      ? relativeSlug
+      : `${relativeSlug}${posix.extname(reference.path)}`;
+  return from(reference, {
+    path: posix.join(posix.dirname(reference.path), slug),
+  });
+};
+
+export const file = (path: string): URI => {
+  let authority = _empty;
+
+  // normalize to fwd-slashes on windows,
+  // on other systems bwd-slashes are valid
+  // filename character, eg /f\oo/ba\r.txt
+  if (isWindows) {
+    path = path.replace(/\\/g, _slash);
+  }
+
+  // check for authority as used in UNC shares
+  // or use the path as given
+  if (path[0] === _slash && path[1] === _slash) {
+    const idx = path.indexOf(_slash, 2);
+    if (idx === -1) {
+      authority = path.substring(2);
+      path = _slash;
+    } else {
+      authority = path.substring(2, idx);
+      path = path.substring(idx) || _slash;
+    }
+  }
+
+  return create('file', authority, path, _empty, _empty);
+};
+
+export const placeholder = (key: string): URI => {
+  return {
+    scheme: 'placeholder',
+    authority: '',
+    path: key,
+    query: '',
+    fragment: '',
+  };
+};
+
+export const from = (
+  uri: URI,
+  change?: {
+    scheme?: string | null;
+    authority?: string | null;
+    path?: string | null;
+    query?: string | null;
+    fragment?: string | null;
+  }
+): URI => {
+  change = change ?? {};
+  const scheme =
+    change.scheme === undefined
+      ? uri.scheme
+      : change.scheme === null
+      ? _empty
+      : change.scheme;
+
+  const authority =
+    change.authority === undefined
+      ? uri.authority
+      : change.authority === null
+      ? _empty
+      : change.authority;
+
+  const path =
+    change.path === undefined
+      ? uri.path
+      : change.path === null
+      ? _empty
+      : change.path;
+
+  const query =
+    change.query === undefined
+      ? uri.query
+      : change.query === null
+      ? _empty
+      : change.query;
+
+  const fragment =
+    change.fragment === undefined
+      ? uri.fragment
+      : change.fragment === null
+      ? _empty
+      : change.fragment;
+
+  return { scheme, authority, path, query, fragment };
+};
+
+export const computeUrisRelativePath = (source: URI, target: URI): string => {
+  const relativePath = posix.relative(posix.dirname(source.path), target.path);
+  return relativePath;
+};
+
+export const getBasename = (uri: URI) => posix.parse(uri.path).name;
+
+export const getDir = (uri: URI) => file(posix.dirname(uri.path));
+
+/**
+ * Uses a placeholder URI, and a reference directory, to generate
+ * the URI of the corresponding resource
+ *
+ * @param placeholderUri the placeholder URI
+ * @param basedir the dir to be used as reference
+ * @returns the target resource URI
+ */
+export const createResourceUriFromPlaceholder = (
```

Comment:
Anything that deals with Foam-specific concepts like placeholders should likely not live inside URI?